src/ray/common/scheduling/scheduling_alg.cc

```diff
@@ -0,0 +1,205 @@
+#include "scheduling_alg.h"
+using namespace std;
+
+ClusterResourceScheduler::ClusterResourceScheduler(int64_t local_node_id, const NodeResources &local_node_resources) {
+  local_node_id_ = local_node_id;
+  AddOrUpdateNode(local_node_id, local_node_resources);
+}
+
+void ClusterResourceScheduler::UpdatePredefinedResources(NodeResources& old_resources,
+  const NodeResources& new_resources) {
+
+    // First, update its predefined resources.
+    for (int i = 0; i < PredefinedResources_MAX; i++) {
+      old_resources.capacities[i].total = new_resources.capacities[i].total;
+      old_resources.capacities[i].available = new_resources.capacities[i].available;
+    }
+}
+
+void ClusterResourceScheduler::UpdateCustomResources(
```

Comment:
We should probably call this `SetCustomResources` actually, as it completely overwrites them